usr/lib/pkcs11/common/attributes.c

```diff
@@ -34,8 +35,10 @@ void free_attribute_array(CK_ATTRIBUTE_PTR attrs, CK_ULONG attrs_len)
         return;
 
     for (i = 0; i < attrs_len; i++)
-        if (attrs[i].pValue)
+        if (attrs[i].pValue) {
+            OPENSSL_cleanse(attrs[i].pValue, attrs[i].ulValueLen);
```

Comment:
not sure about this one: cleansing all attributes per default, sensitive or not. also function name does not indicate memory is cleared before freed. Depends on the context from which this is called. If we can manage to cleanse all sensitive attrs before and *afterwards* call free_attribute_array ?

Reply:
The reason for this is that some sensitive attributes got cpoied into a new array of attributes. 

I have now added a new function free_and_cleanse_attribute_array that can optionally cleanse the attributes before freeing them